src/org/pentaho/reporting/platform/plugin/cache/FileSystemCacheBackend.java

```diff
@@ -51,64 +54,89 @@ public void setCachePath( final String cachePath ) {
 
   @Override
   public boolean write( final List<String> key, final Serializable value ) {
-    final File file = new File( cachePath + StringUtils.join( cleanKey( key ), File.separator ) );
-
-    final ObjectOutputStream oos;
-    final FileOutputStream fout;
-    try {
-      file.getParentFile().mkdirs();
-      if ( !file.exists() ) {
-        file.createNewFile();
+    final List<String> cleanKey = cleanKey( key );
+    synchronized ( getLock( cleanKey ) ) {
+      final String filePath = cachePath + StringUtils.join( cleanKey, File.separator );
+      final File file = new File( filePath );
+      try {
+        //create file structure
+        file.getParentFile().mkdirs();
+        if ( !file.exists() ) {
+          file.createNewFile();
+        }
+        //closable resources
+        try ( final FileOutputStream fout = new FileOutputStream( file );
+              final ObjectOutputStream oos = new ObjectOutputStream( fout ) ) {
+          oos.writeObject( value );
+        }
+      } catch ( final IOException e ) {
+        logger.error( "Can't write cache: ", e );
+        return false;
       }
-
-      fout = new FileOutputStream( file );
-      oos = new ObjectOutputStream( fout );
-      oos.writeObject( value );
-      oos.close();
-      fout.close();
-    } catch ( final IOException e ) {
-      logger.error( "Can't write cache: ", e );
-      return false;
+      return true;
     }
-
-    return true;
   }
 
+
   @Override
   public Serializable read( final List<String> key ) {
-    final ObjectInputStream objectinputstream;
     Object result = null;
-    try {
-      final FileInputStream fis =
-        new FileInputStream( cachePath + StringUtils.join( cleanKey( key ), File.separator ) );
-      objectinputstream = new ObjectInputStream( fis );
-      result = objectinputstream.readObject();
-      objectinputstream.close();
-      fis.close();
-    } catch ( final Exception e ) {
-      logger.debug( "Can't read cache: ", e );
+    final List<String> cleanKey = cleanKey( key );
+    synchronized ( getLock( cleanKey ) ) {
+      final String filePath = cachePath + StringUtils.join( cleanKey, File.separator );
+
+      try ( final FileInputStream fis = new FileInputStream( filePath );
+            final ObjectInputStream ois = new ObjectInputStream( fis ) ) {
+        result = ois.readObject();
+      } catch ( final Exception e ) {
+        logger.debug( "Can't read cache: ", e );
+      }
+      return (Serializable) result;
+    }
+  }
+
+  /**
+   * Returns an object for read/write synchronization
+   * @param key compound key
+   * @return lock object
+   */
+  private synchronized Object getLock( final List<String> key ) {
```

Comment:
No need to sync, no need to check if the "lock" returned is null. 

Your "putIfAbsent" call will never return null. If the map does not contain the lock, it will add the object you have as second parameter to the map and then return it to you. If there is already an lock in the map, it will return that, and will discard your "new Object()".